### PR TITLE
PID-Drive

### PR DESCRIPTION
**Pull request by NoahRoseLedesma**
# PID assisted autonomous driving
## Functionality
- Provides an object that controls autonomous driving
- Corrects errors in robot orientation using a [PID feedback loop](https://en.wikipedia.org/wiki/PID_controller)
## Changes
- [robot.java](TeamCode\src\main\java\org\firstinspires\ftc\teamcode\robot.java) Main source file
  * `robot` Provides object with public drive methods
  * `robotData` Provides robot class with a JSON style organization of members
  * `Drive` Subclass of robotData, contains members retaining to the robot hardware
  * `PID` Subclass of robotData, contains members retaining to the calculation of PID
  * `RobotTime` Subclass of robotData, contains members and methods involving the time of the robot
- [exampleAutonomous.java](TeamCode\src\main\java\org\firstinspires\ftc\teamcode\exampleAutonomous.java) example op-mode implementation of robot object
- [tunePid.java](TeamCode\src\main\java\org\firstinspires\ftc\teamcode\tunePID.java) op-mode for tuning PID values.

### CODE REVIEW

**Comment by NoahRoseLedesma** [comment]
To resolve #3

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/exampleAutonomous.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package org.firstinspires.ftc.teamcode;
+
+import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
+import com.qualcomm.robotcore.hardware.DcMotor;
+import com.qualcomm.robotcore.eventloop.opmode.Autonomous;
+import com.qualcomm.robotcore.hardware.I2cDeviceSynch;
+
+
+@Autonomous(name = "exampleAutonomous", group = "Autonomous")
+public class exampleAutonomous extends LinearOpMode {
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/exampleAutonomous.java]: Please add `@Disabled`

Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/exampleAutonomous.java]: Class names should be `UpperCammelCase`

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+import com.qualcomm.hardware.adafruit.AdafruitBNO055IMU;
+
+import java.util.ArrayList;
+
+/**
+ * Created by Noah Rose Ledesma on 10/1/16. Thanks to Matt Kelsey, Adam Perlin, and Samin Zachariah
+ * Re-implementation of the 2856 PID Driving System without the spaghetti.
+ * Class name is not finalized. Method names are eh right now, looking to revise them.
+ */
+public class robot {
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Class names should be `UpperCammelCase`

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+	public RobotData Data = new RobotData();
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Member names should be `cammelCase`

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+		// Start the program clock
+		Data.Time = new RobotTime();
+	}
+
+	// Public Interface Methods:
+
+	// Method that moves the robot forward variable number of Rotations. Orientation is verified and
+	// corrected by PID control.
+	public void Straight(float Rotations, int Timeout){
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Method names should be `cammelCase`

Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Parameters should be `cammelCase`

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+		// We need two points of data from the IMU to do our calculation. So lets take the first one
+		// and put it into our "current" headings slot.
+		CalculateAngles();
+
+		// Get the current program time and starting encoder position before we start our drive loop
+		float StartTime = Data.Time.CurrentTime();
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Variables should be `cammelCase`.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+		float StartPosition = Data.Drive.leftDrive.getCurrentPosition();
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Variables should be `cammelCase`.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+		Data.PID.IntegralData.clear();
+		Data.PID.DerivativeData.clear();
+
+
+
+		// Manually calculate our first target
+		Data.PID.Target = CalculateAngles() + (IMURotations * 360);
+
+		// We need to keep track of how much time passes between a loop.
+		float LoopTime = Data.Time.CurrentTime();
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Variables should be `cammelCase`.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+			// Record the time since the previous loop.
+			LoopTime = Data.Time.TimeFrom(LoopTime);
+			// Calculate our angles. This method may modify the input Rotations.
+			//IMURotations =
+			CalculateAngles();
+			// Calculate our PID
+			CalculatePID(LoopTime);
+
+			// Calculate the Direction to travel to correct any rotational errors.
+			float Direction = ((Data.PID.I * Data.PID.ITuning) / 2000) + ((Data.PID.P * Data.PID.PTuning) / 2000) + ((Data.PID.D * Data.PID.DTuning) / 2000);
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Variables should be `cammelCase`.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+			// Constrain our direction from being too intense.
+
+			//if(Direction > 50){ Direction = 50; }
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Why is this disabled? Please either remove or explain

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+
+		// Manually calculate our first target
+		Data.PID.Target = CalculateAngles() + (IMURotations * 360);
+
+		// We need to keep track of how much time passes between a loop.
+		float LoopTime = Data.Time.CurrentTime();
+
+		// This is the main loop of our straight drive.
+		// We use encoders to form a loop that corrects rotation until we reach our target.
+		while(Math.abs(StartPosition - Data.Drive.leftDrive.getCurrentPosition()) < Math.abs(Rotations) * Data.Drive.EncoderCount){
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Why do we only use the left motor here?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+			// We are moving forwards
+			Data.Drive.rightDrive.setPower(Drive.POWER_CONSTANT - (Direction));
+			Data.Drive.leftDrive.setPower(Drive.POWER_CONSTANT + (Direction));
+		}
+		// Our drive loop has completed! Stop the motors.
+		Data.Drive.rightDrive.setPower(0);
+		Data.Drive.leftDrive.setPower(0);
+	}
+
+	public void AngleTurn(float angle, int Timeout){
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Methods should be `cammelCase`.

Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Parameters should be `cammelCase`.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+		// We need two points of data from the IMU to do our calculation. So lets take the first one
+		// and put it into our "current" headings slot.
+		CalculateAngles();
+
+		// Get the current program time and starting encoder position before we start our drive loop
+		float StartTime = Data.Time.CurrentTime();
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Variables should be `cammelCase`.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+		float StartPosition = Data.Drive.leftDrive.getCurrentPosition();
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Variables should be `cammelCase`.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+
+		// Reset our Integral and Derivative data.
+		Data.PID.IntegralData.clear();
+		Data.PID.DerivativeData.clear();
+
+
+		//Calculate PIDS again because Isaac Zinda only knows
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Please put a space between `//` and `C`

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+
+
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Remove excess newlines

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+		// Manually calculate our first target
+		Data.PID.Target = (CalculateAngles() + (IMURotations * 360)) + angle;
+
+		// We need to keep track of how much time passes between a loop.
+		float LoopTime = Data.Time.CurrentTime();
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Variables should be `cammelCase`.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+			// Record the time since the previous loop.
+			LoopTime = Data.Time.TimeFrom(LoopTime);
+			// Calculate our angles. This method may modify the input Rotations.
+			//IMURotations =
+			CalculateAngles();
+			// Calculate our PID
+			CalculatePID(LoopTime);
+
+			// Calculate the Direction to travel to correct any rotational errors.
+			float Direction = ((Data.PID.I * Data.PID.ITuning) / 2000) + ((Data.PID.P * Data.PID.PTuning) / 2000) + ((Data.PID.D * Data.PID.DTuning) / 2000);;
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Variables should be `cammelCase`.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+		}
+		// Our drive loop has completed! Stop the motors.
+		Data.Drive.rightDrive.setPower(0);
+		Data.Drive.leftDrive.setPower(0);
+	}
+
+	// Private Methods
+
+	// Method that grabs the IMU data and calculates a new ComputedTarget.
+	private float CalculateAngles(){
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Methods should be `cammelCase`.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+			CalculateAngles();
+		} else if(Data.PID.Headings[0] < Math.abs(60) && Data.PID.Headings[1] > Math.abs(300)) {
+			IMURotations--;
+			CalculateAngles();
+		}
+		return Data.PID.Headings[1];
+	}
+
+	// Method that calculates P, I, and D. Requires the time
+	private void CalculatePID(float LoopTime){
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Methods should be `cammelCase`.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 20]
```diff
@@ -0,0 +1,279 @@
...
+
+
+		// Append to our data sets.
+		Data.PID.IntegralData.add(Data.PID.ComputedTarget - Data.PID.Target);
+		Data.PID.DerivativeData.add(Data.PID.ComputedTarget);
+
+		// Keep IntegralData and DerivativeData from having an exceeding number of entries.
+		if (Data.PID.IntegralData.size() > 500){
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Please move this magic number to a constant. ([Explanation](http://stackoverflow.com/a/47902/1928484))

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 21]
```diff
@@ -0,0 +1,279 @@
...
+			Data.PID.IntegralData.remove(0);
+		}
+
+		if(Data.PID.DerivativeData.size() > 5){
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Please move this magic number to a constant. ([Explanation](http://stackoverflow.com/a/47902/1928484))

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 22]
```diff
@@ -0,0 +1,279 @@
...
+			Data.PID.DerivativeData.remove(0);
+		}
+
+		// Set our P, I, and D values.
+		// `P` will be the ComputedTarget - Target
+		Data.PID.P = Data.PID.ComputedTarget - Data.PID.Target;
+
+		// `I` will be the average of the IntegralData (Cries softly at the lack of Java8 streams)
+		float IntegralAverage = 0;
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Variables should be `cammelCase`.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 23]
```diff
@@ -0,0 +1,279 @@
...
+		for(float value : Data.PID.IntegralData){
+			IntegralAverage += value;
+		}
+		Data.PID.I = IntegralAverage / Data.PID.IntegralData.size();
+
+		// `D` will be the difference of the ComputedTarget and the Derivative average divided by
+		// the time since the last loop in seconds multiplied by one plus half of the size of
+		// the Derivative data set size.
+		float DerivativeAverage = 0;
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Variables should be `cammelCase`.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 24]
```diff
@@ -0,0 +1,279 @@
...
+		Data.PID.D = (Data.PID.ComputedTarget - DerivativeAverage) / ((LoopTime/1000) * (1 + (Data.PID.DerivativeData.size() / 2)));
+	}
+}
+
+
+// Data container classes
+// RobotData acts as the main container for Data.
+// The PID, RobotTime, and Drive Classes act as child data containers for neater organization.
+
+class RobotData {
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Can we just make these members of  `Robot`? Having a bunch of subclasses that are basically structs seems messy to me.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 25]
```diff
@@ -0,0 +1,279 @@
...
+	Drive Drive;
+	RobotData(){
+		PID = new PID();
+		Time = new RobotTime();
+		Drive = new Drive();
+	}
+}
+
+// PID data
+class PID {
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Can we just make these members of  `Robot`? Having a bunch of subclasses that are basically structs seems messy to me.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 26]
```diff
@@ -0,0 +1,279 @@
...
+	ArrayList<Float> IntegralData;
+	// Constructor
+	PID(){
+		// Init non-primitives
+		DerivativeData = new ArrayList<>();
+		IntegralData = new ArrayList<>();
+	}
+}
+// Time data
+class RobotTime {
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Why is this separate from just `ElapsedTime`?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 27]
```diff
@@ -0,0 +1,279 @@
...
+	public float CurrentTime(){
+		return (float) ProgramTime.seconds();
+	}
+
+	public float TimeFrom(float PreviousTime){
+		return (float) (ProgramTime.seconds() - PreviousTime);
+	}
+}
+// Robot hardware data.
+class Drive {
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Can we just make these members of  `Robot`? Having a bunch of subclasses that are basically structs seems messy to me.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java [hunk 28]
```diff
@@ -0,0 +1,279 @@
...
+	//motors indexing around the robot like the quadrants in a graph or like the motors on a drone
+	// for example
+	DcMotor rightDrive;
+	DcMotor leftDrive;
+
+	int EncoderCount;
+	final static float POWER_CONSTANT = (3/8f); // I believe this value does not change. 0.5*(3/4)
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Can we make this a member of `Robot`? I'd like to have all the constants in one place

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/tunePID.java [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
+import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
+import com.qualcomm.robotcore.hardware.DcMotor;
+import com.qualcomm.robotcore.hardware.DcMotorSimple;
+import com.qualcomm.robotcore.hardware.I2cDeviceSynch;
+
+import org.firstinspires.ftc.robotcore.external.Telemetry;
+
+@TeleOp(name = "Tune PID", group = "Util")
+public class tunePID extends LinearOpMode {
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/tunePID.java]: Classes should be `UpperCammelCase`.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java (hunk 4 above)
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Is there any way to merge some of the shared code with `AngleTurn` below?

Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Can we have an overload of this with a default timeout? (Say 1 sec/rotation?)

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java (hunk 11 above)
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: Can we have an overload of this with a default timeout? (Say 1 sec/rotation?)

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/exampleAutonomous.java (hunk 1 above)
Comment (NoahRoseLedesma) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/exampleAutonomous.java]: What is this for?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java (hunk 27 above)
Comment (NoahRoseLedesma) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot.java]: And have a mess of global variables?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Robot.java [hunk 30]
```diff
@@ -0,0 +1,280 @@
...
+		data.PID.D = (data.PID.computedTarget - DerivativeAverage) / ((LoopTime/1000) * (1 + (data.PID.derivativeData.size() / 2)));
+	}
+}
+
+
+// data container classes
+// RobotData acts as the main container for data.
+// The PID, RobotTime, and drive Classes act as child data containers for neater organization.
+
+class RobotData {
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Robot.java]: Can we convert these to members of `Robot`? Having a bunch of structs just seems messy.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Robot.java [hunk 31]
```diff
@@ -0,0 +1,280 @@
...
+			// We are moving forwards
+			data.drive.rightDrive.setPower(data.drive.POWER_CONSTANT - (Direction));
+			data.drive.leftDrive.setPower(data.drive.POWER_CONSTANT + (Direction));
+		}
+		// Our drive loop has completed! Stop the motors.
+		data.drive.rightDrive.setPower(0);
+		data.drive.leftDrive.setPower(0);
+	}
+
+	public void AngleTurn(float angle, int Timeout){
```
Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Robot.java]: Classes should be `UpperCammelCase`.

Comment (ariporad) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Robot.java]: Parameters should be `cammelCase`.

**Comment by ariporad** [comment]
@NoahRoseLedesma: FIX YOUR CODE!

**Pull request merged**